modules/core/src/main/java/org/apache/ignite/internal/tostring/IgniteLogThrottle.java

```diff
@@ -0,0 +1,532 @@
+        log(log, e, null, msg, LogLevel.DEBUG, params);
+    }
+
+    /**
+     * Logs debug if needed.
+     *
+     * @param log Logger.
+     * @param throttleKey Messages with the same key will be throttled.
+     * @param msg Message.
+     */
+    public static void debug(IgniteLogger log, String throttleKey, String msg) {
+        debug(log, throttleKey, msg, null, OBJECT_EMPTY_ARRAY);
+    }
+
+    /**
+     * Logs debug if needed.
+     *
+     * @param log Logger.
+     * @param throttleKey Messages with the same key will be throttled.
+     * @param e Error (optional).
+     * @param msg Message.
+     */
+    public static void debug(IgniteLogger log, String throttleKey, String msg, @Nullable Throwable e) {
+        debug(log, throttleKey, msg, e, OBJECT_EMPTY_ARRAY);
+    }
+
+    /**
+     * Logs debug if needed.
+     *
+     * @param log Logger.
+     * @param throttleKey Messages with the same key will be throttled.
+     * @param msg Message pattern.
+     * @param params List of arguments to be substituted in place of formatting anchors.
+     */
+    public static void debug(IgniteLogger log, String throttleKey, String msg, Object... params) {
+        debug(log, throttleKey, msg, null, params);
+    }
+
+    /**
+     * Logs debug if needed.
+     *
+     * @param log Logger.
+     * @param throttleKey Messages with the same key will be throttled.
+     * @param msg Message pattern.
+     * @param e Error (optional).
+     * @param params List of arguments to be substituted in place of formatting anchors.
+     */
+    public static void debug(IgniteLogger log, String throttleKey, String msg, @Nullable Throwable e, Object... params) {
+        log(log, e, throttleKey, msg, LogLevel.DEBUG, params);
+    }
+
+    /** Clears cached messages. */
+    public static void clean() {
+        MESSAGES_MAP.clear();
+    }
+
+    /**
+     * Logs message if needed using desired level.
+     *
+     * @param log Logger.
+     * @param e Error (optional).
+     * @param longMsg Long message/message pattern.
+     * @param throttleKey Throttle key (optional).
+     * @param level Level where messages should appear.
+     * @param params List of arguments to be substituted in place of formatting anchors.
+     */
+    private static void log(
+            IgniteLogger log,
+            @Nullable Throwable e,
+            @Nullable String throttleKey,
+            String longMsg,
+            LogLevel level,
+            Object... params
+    ) {
+        String errorMessage = nullOrEmpty(params) ? longMsg : IgniteStringFormatter.format(longMsg, params);
+
+        var msgKey = new LogThrottleKey(e, throttleKey == null ? errorMessage : throttleKey);
+
+        while (true) {
+            Long loggedTs = MESSAGES_MAP.get(msgKey);
+
+            long curTs = FastTimestamps.coarseCurrentTimeMillis();
+
+            if (loggedTs == null || loggedTs < curTs - THROTTLE_TIMEOUT_MILLIS) {
```

Comment:
Minor nitpick: I think `curTs - loggedTs > THROTTLE_TIMEOUT_MILLIS` is easier to understand